File: find-pair-given-difference1559.java
/* Given an array, arr[] and an integer x, return true if there exists a pair of elements in the array whose absolute difference is x, otherwise, return false. 
Constraints:
1<= arr.size() <=106 
1<= arr[i] <=106 
0<= x <=105 */


class Solution {
    public boolean findPair(int[] arr, int x) {
        // Sort the array
        Arrays.sort(arr);

        int i = 0;    // First pointer
        int j = 1;    // Second pointer

        x = Math.abs(x);  // Absolute value of the difference

        while (i < arr.length && j < arr.length) {
            int diff = arr[j] - arr[i];

            // If difference matches and i != j
            if (diff == x && i != j) {
                return true;
            }

            // If difference is less than x, move j forward
            if (diff < x) {
                j++;
            }
            // If difference is greater than x, move i forward
            else {
                i++;
            }

            // Ensure pointers do not overlap
            if (i == j) {
                j++;
            }
        }
        return false;
    }
}

File: find-pair-given-difference1559.py
'''
Given an array, arr[] and an integer x, return true if there exists a pair of elements in the array whose absolute difference is x, otherwise, return false.
Constraints:
1<= arr.size() <=106 
1<= arr[i] <=106 
0<= x <=105
'''

from typing import List
class Solution:
    def findPair(self, arr: List[int], x: int) -> int:
        seen = set()
    
        for num in arr:
        # Check if num + x or num - x is in the set
            if num + x in seen or num - x in seen:
                return True
        
        # Add the current element to the set for future comparisons
            seen.add(num)
    
    # If no pair is found
        return False
      
